pkg/syncserver/sync_server.go

```diff
@@ -199,20 +226,43 @@ func (s *Server) SetMaxConns(numConns int) {
 	s.maxConnsC <- numConns
 }
 
+func (s *Server) Port() int {
+	<-s.listeningC
+	return s.chosenPort
+}
+
 func (s *Server) serve(cxt context.Context) {
-	logCxt := log.WithField("port", syncproto.DefaultPort)
+	defer s.Finished.Done()
+	var cancelFn context.CancelFunc
+	cxt, cancelFn = context.WithCancel(cxt)
+	defer cancelFn()
+
+	logCxt := log.WithField("port", s.config.Port)
 	logCxt.Info("Opening listen socket")
-	l, err := net.ListenTCP("tcp", &net.TCPAddr{Port: syncproto.DefaultPort})
+	l, err := net.ListenTCP("tcp", &net.TCPAddr{Port: s.config.ListenPort()})
 	if err != nil {
 		logCxt.WithError(err).Panic("Failed to open listen socket")
 	}
 	logCxt.Info("Opened listen socket")
-	defer l.Close()
+	s.Finished.Add(1)
+	go func() {
+		<-cxt.Done()
+		l.Close()
+		s.Finished.Done()
+	}()
+	s.chosenPort = l.Addr().(*net.TCPAddr).Port
+	logCxt = log.WithField("port", s.chosenPort)
+	close(s.listeningC)
```

Comment:
But my concern was that 'before' might not be well defined for writing a variable on one thread and reading it on another.  I'm pretty sure that if you do this in C, the read thread is not as guaranteed as you might think to get the new value that the other thread wrote - I think because C compilers can reorder lines of code, but perhaps also because the writing thread might only have written to its CPU-local cache, and not yet to the main memory that the other thread is reading from?

(https://en.wikipedia.org/wiki/Memory_barrier)

But perhaps this kind of thing is all safer and properly guaranteed in golang?